src/language-js/printer-estree.js

```diff
@@ -3733,33 +3734,53 @@ function printPropertyKey(path, options, print) {
       (prop) =>
         !prop.computed &&
         prop.key &&
-        isStringLiteral(prop.key) &&
-        !isStringPropSafeToCoerceToIdentifier(prop, options)
+        ((isStringLiteral(prop.key) &&
+          !isStringPropSafeToUnquote(prop, options)) ||
+          (isNumericLiteral(prop.key) &&
+            (options.parser === "flow" ||
+              options.parser === "babel-flow" ||
+              options.parser === "typescript" ||
+              options.parser === "babel-ts")))
```

Comment:
This condition looks wrong.

**Prettier pr-8508**
[Playground link](https://deploy-preview-8508--prettier.netlify.app/playground/#N4Igxg9gdgLgprEAuEBDAvMAjE1AaVXAXxDxAgAcYBLaAZ2VFQCdmIB3ABRYQZVQA27VAE8GZAEbNUYANZwYAZQozqUAObIYzAK5wycALYS4AE1NmAMqg07U6uADEIzQ6hg0NyNDpgRSIAAWMIYCAOqB1PB0KmBwirxR1ABuUSLeYHTiIGp0cMwwnNLqbsgAZoJ5ZABWdAAeAELScgqKqIZwlmpw5ZX6ILV1imrqAnAAijoQ8L0CVSAqzHnM3jAiFHB0YMzUVAEUO7Bh1KYwgcgAHAAMZAcQeWHSFN4Hm-nJPWQAjlPwRZR8cD0ah0eCIMjMOA-aiQor2UpICpzfp5QzULS6FEjMaTaY9RF9MgwVASY6nc5IABMROk1AEIwAwhBDAiQJsAKwBHR5AAqJL4SPmyT0AEkoBZYIptrsYABBcWKNZjWZ5IhEIA)
```sh
--parser typescript
--quote-props consistent
```

**Input:**
```tsx
a={1:a,a:a}
```

**Output:**
```tsx
a = { "1": a, "a": a };

```